### DISCUSSION TITLE
text sizing for vertical vs horizontal images

### DISCUSSION THREAD

**Discussion by DocDJ** [ImageMagick/ImageMagick#3641] [Help]
I have a collection of images, some are vertical, some horizontal. I would like my text "sized" such that the height of each line is the same, regardless of whether the image is vertical or horizontal. I tried using -pointsize to do this, but my verticals wind up with smaller text (top to bottom) than on the horizontal images, even though I specified a -pointsize larger (90 vs 75) for the vertical images. How can I correct for the different orientations?

**Comment by fmw42** [comment]
Please post examples and your exact commands to create them.

**Reply by DocDJ** [reply]
All metadata fields are pre-extracted from each image by the batch file that calls the code below, which is in a separate batch file. The annotations in the resulting images have varying heights depending on whether they are vertical or horizontal and those sizes do not seem to relate to the fontsize. I have also tried replacing `-fontsize` with `-size` and cannot seem to find numeric values to get annotations that are big enough to see.

```
IF %_Height% GTR %_Width% (
	set pointsize=90
	set /A _Yvaltemp= !_Height! * 90
    set /A _Yval= !_Yvaltemp! / 100
) ELSE (
	set pointsize=90
    set /A _Yvaltemp= !_Height! * 85
    set /A _Yval= !_Yvaltemp! / 100
)
magick mogrify -format jpg  -font "Rockwell" -pointsize %pointsize% -weight ExtraBold -fill gold ^
  -stroke black -strokewidth 15 -annotate +45+%_Yval% "%_desc%\n%_country%/%_city%\n%_subloc%" ^
  -stroke none                  -annotate +45+%_Yval% "%_desc%\n%_country%/%_city%\n%_subloc%" ^
  -path "%outpath%" %infile%
```

**Comment by snibgo** [comment]
Please show sample input and output image files: one horizontal, and one vertical.

I suggest you also echo the value of %pointsize%. If your code is contained in a loop, you probably need !pointsize! instead.

**Reply by DocDJ** [reply]
The code I posted is NOT inside a loop. It is passed the values as parameters and DO echo the proper values.
Here are 3 inputs:
![_DSC8564](https://user-images.githubusercontent.com/19713277/118185570-aff9fa00-b40a-11eb-8d05-21f9326849c2.jpg)
![_DSC8572](https://user-images.githubusercontent.com/19713277/118185606-b8eacb80-b40a-11eb-8641-fb3397b31d88.jpg)
![_DSC8588](https://user-images.githubusercontent.com/19713277/118185633-c011d980-b40a-11eb-8d49-3f4178c7abd5.jpg)
Here are 3  outputs:
![_DSC8588](https://user-images.githubusercontent.com/19713277/118186000-31ea2300-b40b-11eb-948a-cc9cf70f9c41.jpg)
![_DSC8564](https://user-images.githubusercontent.com/19713277/118186016-36164080-b40b-11eb-8236-52210a447e6e.jpg)
![_DSC8572](https://user-images.githubusercontent.com/19713277/118186024-39a9c780-b40b-11eb-9dda-152b610f7c97.jpg)

**Comment by snibgo** [comment]
In your three outputs, the text is the same size. The yellow of the capital S is about 64 pixels high.

**Reply by DocDJ** [reply]
if I use a ruler to measure the text in H vs V, the V text is much taller (disregarding the capital letters), even though the specified pointsize is 90 for both. I read the discussion of pointsize in the documentation and it seems the program does not treat it as a constant, being related to the image dimensions. I want to know if there is a way to apply constant text size, which CAN be done with a text IMAGE overlay.

**Comment by snibgo** [comment]
> if I use a ruler to measure the text in H vs V, the V text is much taller ...

Are you displaying the images at the same size? I suspect you are not. I suspect you are displaying them at different sizes.

**Reply by DocDJ** [reply]
same monitor, full screen mode, same application (Windows Photos). I get the same results with a digital photoframe.

**Comment by fmw42** [comment]
Just use -annotate with a given pointsize

**Comment by snibgo** [comment]
When you display the three images, what are their widths and heights, inches?

**Reply by fmw42** [reply]
Going along with snibgo's comment, is it possible that one image is being resized for display at a different ratio than the other image?

**Reply by DocDJ** [reply]
Image 1: 13.5x9, image 2: 6x9, image 3: 13.5x9

**Comment by snibgo** [comment]
So, you have different sizes. If they were the same size, the dimensions would be 13.5x9, 9x13.5 and 13.5x9. You would have the same number of inches, but one format swapping the width and height of the other.

This is like taking a sheet of paper, 13.5x9 inches, and rotating it (without changing the size) to become 9x13.5 inches. 

Put it another way: in pixels, the images are 4288x2848, 4288x2848 and 2848x4288.

If you view the images, with the vertical format resized so the height is the same as the height of the horizontal format, the text (like the rest of the image) is shrunk to fit. You can use a larger pointsize to compensate:
```
set pointsize=%%[fx:90*4288/%_Height%]
```

**Reply by DocDJ** [reply]
I'll have to change your formula slightly, because not all images will be 4288 in width, but I get the idea. I'll try it in the next few days and post my results. Thanks for the explanation.

**Reply by DocDJ** [reply]
I tried your formula and it has a quirk. For most of the above images, the text came out as expected, but for the one attached here (1447x2116), it was huge. The other vertical image (2848x4288), previously posted, came out OK. Any ideas what's happening?

![DSC_0317_edited](https://user-images.githubusercontent.com/19713277/119343010-d833f500-bc63-11eb-8983-f3006e6902e5.jpg)

**Comment by snibgo** [comment]
What formula are you using? What is the value of pointsize? What value do you want instead?

My suggestion was based on the assumption that your images were either 4288x2848 or 2848x4288. For other sizes, what happens with the images? Do all your images enlarge or shrink to fit inside a certain box in your picture frame? If so, then your script has to account for that resizing, and adjust the value of pointsize.

**Reply by DocDJ** [reply]
the images here are all pasted directly from my computer (copy/paste directly from the file). I don't understand why the formula doesn't maintain the point sizes for the 2 vertical images. W/H is .66 for one and .68 for the other so they should be at least CLOSE in size, but they are not at all close. I get the same disparity for the horizontal images, but nowhere near as grossly oversized as the vertical (1447x2116).
The formula I'm using is:
` set /A t1=!base!*!_Height!
        set /A relpointsize=!t1!/!_Width!
`
DOS batch doesn't do floating point, so I have to multiply, then divide to get the relative pointsize. base is set to 90.

**Comment by snibgo** [comment]
You don't show values of t1 and relpointsize.

I suspect that the pointsize is as you have specified, but you are viewing the results at different enlargements.

Whether the image is large or small, you are using the same pointsize. But you enlarge small images by a greater factor. Enlarging it more also enlarges the text.

**Reply by DocDJ** [reply]
the code is actually 2 lines (doesn't show  up that way in my reply. t1 is set in the part of what is shown. relpointsize is in the 2nd part. I think I see what you mean. The 2 images are getting expanded differently to take up the vertical space in the viewing window. Since one image is physically smaller, it has to be expanded more and the text is being expanded too, but not proportionally. I thought the annotation was burned INTO the  image so the pixels would be expanded the same amount as the rest of the image.  I guess I need a better formula to allow for the smaller files.

**Comment by snibgo** [comment]
> I thought the annotation was burned INTO the image so the pixels would be expanded the same amount as the rest of the image.

Yes, that is correct. And small images are expanded more than large images. So the text in a small image is expanded more than the text in a large image.

**Reply by DocDJ** [reply]
now it makes sense. Since I plan to display the same images (and annotations) from my website and my photo frame, I guess I will need different formulas for both outputs as well as for different image sizes. Is that correct?

**Comment by snibgo** [comment]
> Is that correct?

It depends.

When you have two images, one small and one large, they are expanded by different amounts for your photo frame, so text expands by different amounts.

That may be true of your website. On the other hand, perhaps your website shows small images as small, and your large images as large, so they expand by the same amount, so text doesn't change in size.

**Reply by DocDJ** [reply]
I'll have to see what my 4 images look like on the website (have to get them there first.) I'll post some results as soon as I have them.

**Reply by DocDJ** [reply]
Please check the images on my website: [https://dforeman.homedns.org/Tpix](https://dforeman.homedns.org/Tpix)

**Comment by snibgo** [comment]
That link is to a list of JPG files with large images. When I view them in Firefox, they are reduced in size, unless I click on the "magnify" icon so the image (including text) is shown with pixels 1:1 on my screen.

And...?

**Reply by DocDJ** [reply]
when I open them in FF, they appear full-sized. In Edge, same but I have to scroll them to see the text.
This is a screen-capture of a horizontal using FF.
![image](https://user-images.githubusercontent.com/19713277/119726707-657f7100-be3f-11eb-9806-819542b8e93a.png)
This is a screen-capture of a vertical using FF.
![image](https://user-images.githubusercontent.com/19713277/119727027-d030ac80-be3f-11eb-9a48-175557a95c37.png)

**Comment by snibgo** [comment]
Those images are resized to be the same height. If you want those results to have the same height of text, then you need to set the pointsize as a proportion of the height of each image, the same proportion for the two images.

**Reply by DocDJ** [reply]
thanks a million. That seems to work pretty well. I'm now experimenting on finding the right multiplier. 1/33 is pretty close.
